First sketch of schemas for computational provenance in EBRAINS

apdavison opened the pull request:
I'm not sure about how the extension mechanism works, so this will almost certainly need work.

apdavison commented:
<img width="1357" alt="image" src="https://user-images.githubusercontent.com/260552/109303135-a06f0a00-783a-11eb-8aba-6f932a3e0d0a.png">

mfahdaz commented:
It would be interesting how some interesting information could be retrieved later using this schema such as 

- Resource utilization patterns for different software of a computations, i.e. how much and which resources were used by a specific computation (In this case, more utilization data might be needed to be stored than just the core hours),
- This could further be decomposed to the specific software of a computation such that how much and which resources were used by a specific software of a computation (this could help later in identifying the most resource hungry component of the cosim framework),
- Correlation between different environment/hardware/parameter settings and the execution time (performance) of a (given software of a) computation, etc.

Could it be possible to also store the power/energy consumption data for computations i.e. how much power is consumed by a given computation?

w-klijn commented:
From the json files,

the resourceUsage type should be array to capture more data than just core hours (that's what I pointed to in the comment).

Further It is not clear whether the computation data is stored for models and software as it has 1 to many relationship with them but the json files does not include any link to them... I'm not sure if the relations are already defined in the schemas (in other repos such as core/research/protocolExecution.schema.tpl.json) which it is extending. So, I just left a comment to mention the case to be covered.

Lastly, based on the current design the model version schema can further be decomposed to a set of simulation actions and plan. I left it however so we could finalize the schemas for that ourselves in tomorrow's meeting before pointing it out here.

apdavison commented:
Michael Denker commented [here](https://gitlab.humanbrainproject.org/joint_infrastructure_coordination/Coordination/-/issues/228#note_23630):

Regarding the Computation component, it may be worth considering adding a log entry that would point to a resource (e.g., a File) that contains the execution logs. Although these log files are in a way part of the output, at the same time they play a very special role in the context of workflow execution and a user may want to be able to quickly access these info.

apdavison commented:
Michael Denker commented [here](https://gitlab.humanbrainproject.org/joint_infrastructure_coordination/Coordination/-/issues/228#note_23631):

Regarding Protocol, I find the term fitting, and it relates to the existing schemas. However, I could imagine some people may not associate an execution call as computational protocol, but rather only an experimental process. One may consider a different name as alias ('Execution', 'Call", "Spawn",...), or a more generic term.

apdavison commented:
@mfahdaz @w-klijn as you suggested I've made "resourceUsage" an array so we can track consumption of different types of resource.

@w-klijn I've added Optimization as a specific sub-class of Computation that can have ModelVersion as output. For now, I've removed the "output" link to SoftwareVersion from the diagram, although we may wish to add a new computation sub-class "CodeGeneration" (for NESTML, etc.)

@mdenker re: log files, I think we could handle this by giving the FileInstance representing the log files a specific content type, but this should be discussed further.

@mdenker re: Protocol. I agree, this is the weakest part of the current proposal; the name comes from the parent class of Computation, core.ProtocolExecution. However, we may consider having an abstract parent class from which both Computation and Protocol Execution will inherit, in which case a more appropriate name and type could be chosen.

UlrikeS91 commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+    "_type": "https://openminds.ebrains.eu/computation/Computation",
+    "_extends": "../../core/research/protocolExecution.schema.tpl.json",
+    "required": [
+        "input",
+        "output",
+        "protocol",
```
You could remove those since they are already required fields for the protocolExecution schema itself.

UlrikeS91 commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+{
+    "_type": "https://openminds.ebrains.eu/computation/Computation",
+    "_extends": "../../core/research/protocolExecution.schema.tpl.json",
```
@lzehl, any idea how to reference this? We never actually extended anything across repos before...

lzehl replied:
We do not support this momentarily, but there are good reasons to support this and it should not be too tricky to adapt the openMINDS generator accordingly. Technical details need to be discussed with Oli.

UlrikeS91 commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+{
+    "_type": "https://openminds.ebrains.eu/computation/Computation",
+    "_extends": "../../core/research/protocolExecution.schema.tpl.json",
```
Additionally, "_extends" will cause your schema (here: computation) to be extended by the content of the ref. schema (here: protocolExecution). 
Therefore, the 3 properties - "input", "output" and "protocol" - shouldn't be listed under properties here. You also don't have to add them as required because the original schema (protocolExecution) has them required already. 
However, looking at the details, I noticed that you would like to change the "_linkedTypes" for "input" and "output" here. As far as I'm aware, this is not possible like this. 
I'm unsure what you would like to achieve here. 
1. Do you want “computation” to have all properties from protocolExecution and the additional ones stated specifically here? Then “computation” would be a more detailed version of a “protocolExecution”. 
OR
2. Do you want to express that this “computation” is related to a specific “protocolExecution”? Then the “protocolExecution” would exist first, and the “computation” gives additional information related to the execution. 
I’m asking to figure out how the schema would need to be technically changed to express the correct alternative.   
For 1.: “input”, “output” and “protocol” are already existing properties for “protocolExecution”. These are required there, so that you could remove them from the required field for “computation”. The properties wouldn’t need to be listed under “properties” again and the original content cannot be overwritten. You used other “_linkedTypes” for “input” and “output” for “computation” than “protocolExecution”. As far as I know that’s not possible/allowed. This would need to be changed then. 
For 2.: “_extends” wouldn’t need to be used at all. Then there should be a new property e.g. “protocolExecution” or “relatedExperiment” with “_linkedTypes” pointing to a “protocolExecution” (I assume as an object rather than an array then).

apdavison replied:
Thanks for the quick feedback Ulrike. I am trying to achieve option 1, but I wasn't sure about the rules for "_extends" in terms of extending or restricting existing fields. One approach would be to define a parent type like "Activity" that could be extended by both ProtocolExecution and Computation. The question of how to extend core types definitely needs to be addressed more generally, though, as I think this will be a common need in openMINDS extensions (ephys, etc.)

UlrikeS91 replied:
Yes, this was what I was afraid of :D Option 2 would have been easier. 

@apdavison, you are right. @lzehl this is something that needs to be discussed. With every additional extension, some existing schemas may need to be split to allow this "recycling mechanism" that `"_extends"` entails.

The easy fix for this specific case is as you suggested, @apdavison. Our "protocolExecution" schema would need to extend a new schema which holds the properties that will be identical for both "protocolExecution" and "computation" - everything except for input, output and protocol. 
However, "computation" would not be able to have the property "input" itself. This would need to be part of the 3 sub-schemas that extend "computation" since the "input" for "simulation" expects other linkedTypes than "visualization"/"dataAnalysis".  


Slightly unrelated: I would personally like to suggest to avoid using a schema name twice. I got confused here thinking that protocol referred to the core/protocol until I noticed that there was a computation/protocol. If it's just me that gets confused by this, please do ignore this suggestion ;)

olinux replied:
As discussed, we should support this for submodules of the central repository. I would like to suggest     
"_extends": "/core/research/protocolExecution.schema.tpl.json",

apdavison commented:
<img width="973" alt="image" src="https://user-images.githubusercontent.com/260552/113340871-b9cd0f80-932c-11eb-93c5-a08911b0de4c.png">

lzehl commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+{
+    "_type": "https://openminds.ebrains.eu/computation/Computation",
+    "_extends": "/core/research/activity.schema.tpl.json",
```
@apdavison : activity.schema.tpl.json as the superclass of Computation and ProtocolExecution?

lzehl replied:
Would activity.schema.tpl.json hold only the following properties?
+ name
+ description
+ input
+ output

apdavison replied:
I'm planning a PR against core to implement this. Probably also parameterSet, studyTarget; possibly semanticallyAnchoredTo, startedAtTime, endedAtTime, wasInformedBy

lzehl commented on schemas/launchconfiguration.schema.tpl.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "_type": "https://openminds.ebrains.eu/computation/LaunchConfiguration",
```
note: we keep lowerCaseCamel for the filenames as well.

apdavison replied:
unclear what this refers to, Github is showing me "_type", which does not reference a filename

lzehl commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+            "_linkedTypes": [
+                "https://openminds.ebrains.eu/core/FileBundle",
+                "https://openminds.ebrains.eu/core/FileInstance",
+                "https://openminds.ebrains.eu/products/SoftwareVersion"
```
How can SoftwareVersion be an input? Is a SoftwareVersion actually manipulated by the Computation to produce a different SoftwareVersion as output or is it used by the Computation to produce an output?

apdavison replied:
This would be the user's own code, typically, e.g. the top-level Python script. 

You could also include all dependencies (e.g. NumPy, ...), but usually these would be considered part of the Environment. If you installed a specific version of a dependency that is different to the version in the environment, it could be added here.

lzehl commented on schemas/optimization.schema.tpl.json:
```diff
@@ -0,0 +1,19 @@
+        "_linkedTypes": [
+            "https://openminds.ebrains.eu/core/FileBundle",
+            "https://openminds.ebrains.eu/core/FileInstance",
+            "https://openminds.ebrains.eu/products/SoftwareVersion",
```
same question as above. How can SoftwareVersion be an input? For ModelVersion: Is it the case that one ModelVersion is an input, is optimized by the computation (here optimization) and then another ModelVersion comes out as output?

apdavison replied:
for SoftwareVersion, answer as above. For ModelVersion, yes.

lzehl commented on schemas/softwareagent.schema.tpl.json:
```diff
@@ -0,0 +1,25 @@
+{
```
note: we use also lowerCamelCase for the file names.

lzehl commented on schemas/workflowexecution.schema.tpl.json:
```diff
@@ -0,0 +1,22 @@
+{
+    "_type": "https://openminds.ebrains.eu/computation/WorkflowExecution",
```
note: we use also lowerCamelCase for the file names.

apdavison replied:
?

lzehl commented on schemas/computation.schema.tpl.json:
```diff
@@ -0,0 +1,84 @@
+    "required": [
+        "input",
+        "output",
+        "protocol",
```
I guess "protocol" needs to be now "launchConfiguration"?

apdavison replied:
yes, thanks. Fixed.

lzehl reviewed (commented):
@apdavison I briefly went through the metadata model and made some comments / had some questions, but I think this is looking good. I guess the controlledTerms could be extended by quite a few terminologies: for the studyTarget (I guess), but also maybe some other attributes where you want to better control the vocabulary?
